File: app/src/main/java/com/example/loren/minesample/VectorActivity.java
package com.example.loren.minesample;

import android.animation.ValueAnimator;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.hanks.htextview.util.DisplayUtils;

import java.util.Locale;

import butterknife.BindView;
import butterknife.ButterKnife;

public class VectorActivity extends AppCompatActivity {

    @BindView(R.id.vector_view)
    TextView vectorView;
    @BindView(R.id.volume_view)
    TextView volumeView;
    @BindView(R.id.volume_container_ll)
    LinearLayout volumeContainerLl;
    @BindView(R.id.cur_volume_tv)
    TextView curVolumeTv;
    private AudioRecordDemo mAudioRecord;
    private int CHANGE_SIZE = 50;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_vector);
        ButterKnife.bind(this);
        for (int i = 0; i < 15; i++) {
            View item = LayoutInflater.from(this).inflate(R.layout.volume_item, null);
            volumeContainerLl.addView(item);
        }
        mAudioRecord = new AudioRecordDemo(new AudioRecordDemo.OnVolumeListener() {
            @Override
            public void onVolumeListener(final double volume) {
                Log.d("loren", "分贝值：" + volume);
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        curVolumeTv.setText("分贝值：" + String.format(Locale.CHINA, "%.2f", volume));
                    }
                });
                for (int i = 0; i < volumeContainerLl.getChildCount(); i++) {
                    final TextView view = (TextView) ((RelativeLayout) volumeContainerLl.getChildAt(i)).getChildAt(0);
                    final float heightNum = (float) (volume * (i > (volumeContainerLl.getChildCount() / 2) ? (volumeContainerLl.getChildCount() - i) : i));
                    addListener(view);
                    view.postDelayed(new Runnable() {
                        @Override
                        public void run() {
                            final ViewGroup.LayoutParams params = view.getLayoutParams();
                            int heightTemp = params.height;
                            ValueAnimator animator = ValueAnimator.ofInt(heightTemp,
                                    DisplayUtils.dp2px(VectorActivity.this, heightNum));
                            animator.setDuration(100);
                            animator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
                                @Override
                                public void onAnimationUpdate(ValueAnimator animation) {
                                    params.height = (int) animation.getAnimatedValue();
                                    view.setLayoutParams(params);
                                }
                            });
                            animator.start();
                        }
                    }, 50 * i);
                }
            }
        });
        mAudioRecord.getNoiseLevel();
    }

    private void addListener(final View view) {
        view.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_UP:
                        changeParams(v, MotionEvent.ACTION_UP);
                        break;
                    case MotionEvent.ACTION_DOWN:
                        changeParams(v, MotionEvent.ACTION_DOWN);
                        break;
                    case MotionEvent.ACTION_MOVE:
                        break;
                }
                return true;
            }
        });
    }

    private void changeParams(View view, int event) {
        ViewGroup.LayoutParams params = view.getLayoutParams();
        if (event == MotionEvent.ACTION_DOWN) {
            params.width += CHANGE_SIZE;
            params.height += CHANGE_SIZE;
        } else {
            params.width -= CHANGE_SIZE;
            params.height -= CHANGE_SIZE;
        }
        view.setLayoutParams(params);
    }

    @Override
    protected void onDestroy() {
        mAudioRecord.stopAudioThread();
        super.onDestroy();
    }
}

File: app/src/main/java/com/example/loren/minesample/VectorActivity.kt
package com.example.loren.minesample

import android.Manifest
import android.animation.ValueAnimator
import android.annotation.SuppressLint
import android.util.Log
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.widget.RelativeLayout
import android.widget.TextView
import com.example.loren.minesample.base.ui.BaseActivity
import kotlinx.android.synthetic.main.activity_vector.*
import pers.victor.ext.dp2px
import pers.victor.ext.toast
import java.util.*


class VectorActivity : BaseActivity() {
    @SuppressLint("SetTextI18n")
    override fun initWidgets() {
        requestPermission(Manifest.permission.RECORD_AUDIO, granted = { }, denied = { toast("没有权限，请授权后重试") })
        for (i in 0..14) {
            val item = LayoutInflater.from(this).inflate(R.layout.volume_item, null)
            volume_container_ll!!.addView(item)
        }
        mAudioRecord = AudioRecordDemo(AudioRecordDemo.OnVolumeListener { volume ->
            Log.d("loren", "分贝值：$volume")
            runOnUiThread { cur_volume_tv!!.text = "分贝值：" + String.format(Locale.CHINA, "%.2f", volume) }
            for (i in 0 until volume_container_ll!!.childCount) {
                val view = (volume_container_ll!!.getChildAt(i) as RelativeLayout).getChildAt(0) as TextView
                val heightNum = (volume * if (i > volume_container_ll!!.childCount / 2) volume_container_ll!!.childCount - i else i).toFloat()
                addListener(view)
                view.postDelayed({
                    val params = view.layoutParams
                    val heightTemp = params.height
                    val animator = ValueAnimator.ofInt(heightTemp, dp2px(heightNum))
                    animator.duration = 100
                    animator.addUpdateListener { animation ->
                        params.height = animation.animatedValue as Int
                        view.layoutParams = params
                    }
                    animator.start()
                }, (50 * i).toLong())
            }
        })
        mAudioRecord!!.getNoiseLevel()
    }

    override fun setListeners() {
    }

    override fun onWidgetsClick(v: View) {
    }

    override fun bindLayout() = R.layout.activity_vector

    private var mAudioRecord: AudioRecordDemo? = null
    private val CHANGE_SIZE = 50

    private fun addListener(view: View) {
        view.setOnTouchListener { v, event ->
            when (event.action) {
                MotionEvent.ACTION_UP -> changeParams(v, MotionEvent.ACTION_UP)
                MotionEvent.ACTION_DOWN -> changeParams(v, MotionEvent.ACTION_DOWN)
                MotionEvent.ACTION_MOVE -> {
                }
            }
            true
        }
    }

    private fun changeParams(view: View, event: Int) {
        val params = view.layoutParams
        if (event == MotionEvent.ACTION_DOWN) {
            params.width += CHANGE_SIZE
            params.height += CHANGE_SIZE
        } else {
            params.width -= CHANGE_SIZE
            params.height -= CHANGE_SIZE
        }
        view.layoutParams = params
    }

    override fun onDestroy() {
        mAudioRecord!!.stopAudioThread()
        super.onDestroy()
    }
}
